cmd/queue.go

```diff
@@ -0,0 +1,382 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"log"
+	"time"
+
+	redisqueue "github.com/frain-dev/convoy/queue/redis"
+	"github.com/frain-dev/convoy/util"
+	"github.com/go-redis/redis/v8"
+	"github.com/spf13/cobra"
+	"github.com/vmihailenco/taskq/v3"
+)
+
+const argsSlice = 12
+
+func addQueueCommand(a *app) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "queue",
+		Short: "Get info about queue",
+	}
+
+	cmd.AddCommand(getQueueLength(a))
+	cmd.AddCommand(getZSETLength(a))
+	cmd.AddCommand(getStreamInfo(a))
+	cmd.AddCommand(getConsumersInfo(a))
+	cmd.AddCommand(getPendingInfo(a))
+	cmd.AddCommand(checkEventDeliveryonQueue(a))
+	cmd.AddCommand(checkEventDeliveryinZSET(a))
+	cmd.AddCommand(checkEventDeliveryinPending(a))
+	return cmd
+}
+
+//Get queue length, number of entries in the stream
+func getQueueLength(a *app) *cobra.Command {
+	var timeInterval int
+	cmd := &cobra.Command{
+		Use:   "length",
+		Short: "queue length",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			q := a.eventQueue
+			ctx := context.Background()
+			ticker := time.NewTicker(time.Duration(timeInterval) * time.Millisecond)
+
+			for {
+				select {
+				case <-ticker.C:
+					length, err := q.Consumer().Queue().Len()
+					if err != nil {
+						log.Printf("Error getting queue length: %v", err)
+					}
+					log.Printf("Queue Length: %+v\n", length)
+				case <-ctx.Done():
+					return nil
+				}
+			}
+		},
+	}
+	cmd.Flags().IntVar(&timeInterval, "interval", 2000, "Log time interval")
+	return cmd
+}
+
+//get length of ZSET, delayed msgs
+func getZSETLength(a *app) *cobra.Command {
+	var timeInterval int
+	cmd := &cobra.Command{
+		Use:   "zsetlength",
+		Short: "get ZSET Length",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			q := a.eventQueue.(*redisqueue.RedisQueue)
```

Comment:
Question, does this technically break if we use another queue client?
It feels like it would.

Do we also just want to encapsulate and expose methods for the range in the interface of the queue itself? as against the implementation assertions here?

Or are we certain that we'd only ever use a redis queue (and then, need observability against the redis queue implementation only)?

Reply:
yes it would break stating that that `a.eventQueue` is not of type `redis`. A command that has `zset` in it is clearly for redis queue only. Also in terms of encapsulating the commands here, taskq already encapsulates these commands behind the RedisStreamClient which is an interface of type `redis.Client ` hence my need to access the redis client through "inner".